backend/modules/metadata_store/prisma_store.py

```diff
@@ -195,183 +224,122 @@ async def aget_data_sources(self) -> List[DataSource]:
     ) -> Collection:
         try:
-            collection: Collection = await self.aget_collection_by_name(collection_name)
-        except Exception as e:
-            logger.exception(f"Error: {e}")
-            raise HTTPException(status_code=500, detail=f"Error: {e}")
-
-        if not collection:
-            logger.error(f"Collection with name {collection_name} does not exist")
-            raise HTTPException(
-                status_code=400,
-                detail=f"Collection with name {collection_name} does not exist",
-            )
-
-        try:
-            data_source = await self.aget_data_source_from_fqn(data_source_fqn)
-        except Exception as e:
-            logger.exception(f"Error: {e}")
-            raise HTTPException(status_code=500, detail=f"Error: {e}")
-
-        if not data_source:
-            logger.error(f"Data source with fqn {data_source_fqn} does not exist")
-            raise HTTPException(
-                status_code=400,
-                detail=f"Data source with fqn {data_source_fqn} does not exist",
-            )
+            # Get the collection by name
+            collection = await self.aget_collection_by_name(collection_name)
+            # Get the existing associated data sources
+            associated_data_sources = collection.associated_data_sources or {}
+            # If the data source is not associated with the collection, deletion is not possible and an error is raised
+            if data_source_fqn not in associated_data_sources:
+                raise HTTPException(
+                    status_code=400,
+                    detail=f"Data source with fqn {data_source_fqn!r} not associated with collection {collection_name!r}",
+                )
+            # Remove the data source from associated data sources
+            associated_data_sources.pop(data_source_fqn)
 
-        associated_data_sources = collection.associated_data_sources
-        if not associated_data_sources:
-            logger.error(
-                f"No associated data sources found for collection {collection_name}"
-            )
-            raise HTTPException(
-                status_code=400,
-                detail=f"No associated data sources found for collection {collection_name}",
-            )
-        if data_source_fqn not in associated_data_sources:
-            logger.error(
-                f"Data source with fqn {data_source_fqn} not associated with collection {collection_name}"
-            )
-            raise HTTPException(
-                status_code=400,
-                detail=f"Data source with fqn {data_source_fqn} not associated with collection {collection_name}",
-            )
-        associated_data_sources.pop(data_source_fqn, None)
+            # Convert the associated data sources to a dictionary
+            updated_associated_data_sources = {
+                fqn: ds.model_dump() for fqn, ds in associated_data_sources.items()
+            }
```

Comment:
See [ref](https://github.com/truefoundry/cognita/pull/361#discussion_r1792462418).

Also, can you give me an example of how joins will help here? I can go ahead accordingly.

Reply:
https://github.com/truefoundry/cognita-internal/blob/a7028b1dabb557f27c7dc8631159917ed306668a/backend/database/schema.prisma#L44-L45

I was doing it this way, then using an include on the user to get the user's collections. Perhaps there is a better way to handle joins?

Reply:
Yeah, like I mentioned previously, we need to figure out the best way to add an association between data sources and collections. It's best if we can tackle this in a separate PR since it will involve migrations, updation of code in various places etc